File: src/my_py_pkg/my_py_pkg/add_2_ints_server.py
#!/usr/bin/env python3
#

import rclpy
from rclpy.node import Node

# ros2 interface show example_interfaces/srv/AddTwoInts
# AddTwoInts.a <- name of the first op. 
# AddTwoInts.b <- name of the second op.
# AddTwoInts.sum <- response sum = a + b.
from example_interfaces.srv import AddTwoInts

SERVICE_NAME = "add_2_ints"

class Add2IntsServerNode( Node ):
    def __init__( self ):
        super().__init__( "add_2_ints_server" )
        self.server_ = self.create_service( AddTwoInts, SERVICE_NAME,
                                            self.callback_add_2_ints )
        self.get_logger().info( "ADD2IntsServer has been started." )

    def callback_add_2_ints( self, request: AddTwoInts.Request,
                             response: AddTwoInts.Response ):
        response.sum = request.a + request.b
        self.get_logger().info( str( request.a ) + " + " + 
                               str( request.b ) +
                               " = " + str( response.sum ) )
        return response
    

def main( args = None ):
    rclpy.init( args = args )
    node = Add2IntsServerNode()
    rclpy.spin( node )
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/my_cpp_pkg/src/add_2_ints_server.cpp
#include "rclcpp/rclcpp.hpp"
/*
# ros2 interface show example_interfaces/srv/AddTwoInts
# AddTwoInts.a <- name of the first op. 
# AddTwoInts.b <- name of the second op.
# AddTwoInts.sum <- response sum = a + b.
*/
#include "example_interfaces/srv/add_two_ints.hpp"

// using namespace std::placeholders;

#define SERVICE_NAME "add_2_ints"

class Add2IntsServerNode : public rclcpp::Node
{
    public:
        Add2IntsServerNode() : Node( "add_2_ints_server" )
        {
            this->server_ = this->create_service<example_interfaces::srv::AddTwoInts>( SERVICE_NAME,
                                                std::bind( &Add2IntsServerNode::callbackAdd2Ints, this,
                                                std::placeholders::_1, std::placeholders::_2 ) );
            RCLCPP_INFO( this->get_logger(), "Add2IntsServer has been started." );
        };

    private:
        rclcpp::Service<example_interfaces::srv::AddTwoInts>::SharedPtr server_;

        void callbackAdd2Ints( const example_interfaces::srv::AddTwoInts::Request::SharedPtr request,
                               const example_interfaces::srv::AddTwoInts::Response::SharedPtr response )
        {
            response->sum = request->a + request->b;
            RCLCPP_INFO( this->get_logger(), "%ld + %ld = %ld",
                         request->a, request->b, response->sum );
        };
};

int main( int argc, char **argv )
{
    rclcpp::init( argc, argv );
    auto node = std::make_shared<Add2IntsServerNode>();
    rclcpp::spin( node );
    rclcpp::shutdown();
    return 0;
};
